### ISSUE TITLE
Failed to add error rates to overlap store

### ISSUE THREAD

**Issue by dyermd** [marbl/canu#59]
I am running Canu on the test PacBio dataset you provide (renamed to reads.fastq) and I keep running into a `failed to add error rates to overlap store` error. The command I am using is 

`canu -p reads -d /data genomeSize=4800000 -pacbio-raw /data/reads.fastq`

Below is tail of the log.

```
----------------------------------------
-- Found 1 read error detection output files.
-- overlap error adjustment attempt 0 begins with 0 finished, and 1 to compute.
----------------------------------------
-- Starting concurrent execution on Tue Mar  1 03:18:52 2016 with 174.8 GB free disk space (1 processes; 1 concurrently)

    /data/unitigging/3-overlapErrorAdjustment/oea.sh 1 > /data/unitigging/3-overlapErrorAdjustment/oea.000001.out 2>&1

-- Finished on Tue Mar  1 03:18:59 2016 (7 seconds) with 174.8 GB free disk space
----------------------------------------
-- Found 1 overlap error adjustment output files.
----------------------------------------
-- Starting command on Tue Mar  1 03:18:59 2016 with 174.8 GB free disk space

    /usr/src/app/canu/Linux-amd64/bin/ovStoreBuild \
      -G /data/unitigging/reads.gkpStore \
      -O /data/unitigging/reads.ovlStore \
      -evalues \
      -L /data/unitigging/3-overlapErrorAdjustment/oea.files \
    > /data/unitigging/3-overlapErrorAdjustment/oea.apply.err 2>&1

-- Finished on Tue Mar  1 03:18:59 2016 (lickety-split) with 174.8 GB free disk space
----------------------------------------
ERROR:
ERROR:  Failed with exit code 1.
ERROR:
================================================================================
Don't panic, but a mostly harmless error occurred and canu failed.

Disk space available:  174.8 GB

Last 50 lines of the relevant log file (/data/unitigging/3-overlapErrorAdjustment/oea.apply.err):


canu failed with 'failed to add error rates to overlap store'.
```

If I run that last command that failed directly by hand I see the following error message

```
loading evalues from '/data/unitigging/3-overlapErrorAdjustment/0001.oea'
loading evalues from '/data/unitigging/3-overlapErrorAdjustment/0001.oea' -- ID range 1-10902 with 399547 overlaps
WARNING: closing read-only evalues file.
memoryMappedFile()-- Couldn't mmap '/data/unitigging/reads.ovlStore/evalues': Invalid argument
```

that `evalues` file is there

```
$ ls
0001	evalues	index	info
```

**Comment by brianwalenz** [comment]
What OS are you on?  Are you in a VM?

In file src/AS_UTL/memoryMappedFile.H, change line 197 from:

```
    _data = (_type == memoryMappedFile_readOnly) ? mmap(0L, _length, PROT_READ,              MAP_FILE | MAP_SHARED, fd, 0)
                                                 : mmap(0L, _length, PROT_READ | PROT_WRITE, MAP_FILE | MAP_SHARED, fd, 0);
```

to:

```
    _data = (_type == memoryMappedFile_readOnly) ? mmap(0L, _length, PROT_READ,              MAP_FILE | MAP_SHARED, fd, 0)
                                                 : mmap(0L, _length, PROT_READ | PROT_WRITE, MAP_FILE | MAP_PRIVATE, fd, 0);
```

That is, MAP_SHARED to MAP_PRIVATE in the second line.

After that "gmake" in src/, remove the 'evalues' file and run the command again.

**Comment by dyermd** [comment]
Thanks for the suggestion. I will give that a shot later today. I am running this from a docker container.

**Comment by dyermd** [comment]
@brianwalenz - just had a chance to try it out and that fixed it. Thanks.

Just a quick note, it was like 97, not 197

**Comment by brianwalenz** [comment]
For the record, MAP_PRIVATE is NOT a proper fix.  It doesn't crash, but the data it is trying to update (on disk) is lost.

**Comment by skoren** [comment]
The issue is MAP_SHARED is not implemented between the VM and host machine. I'm guessing you're running canu on a mounted filesystem from your real machine inside docker (e.g. the -v path). You need to run fully within the VM filesystem for mmap to work. When I ran:

```
docker run -t -i --rm   -v `pwd`:/io   -e CC='gcc'   -e CXX='g++'   phusion/holy-build-box-64:latest   bash
cd /io
/io/canu-1.1/Linux-amd64/bin/canu -p asm -d asm -pacbio-raw /io/test/pacbio.filtered_subreads.fasta genomeSize=50k java=/io/test/jre1.8.0_73/bin/java minReadLength=500 stopOnReadQuality=false

-- Starting command on Sun Mar 13 18:49:08 2016 with 372.2 GB free disk space

    /io/canu/Linux-amd64/bin/ovStoreBuild \
      -G /io/test/asm/unitigging/asm.gkpStore \
      -O /io/test/asm/unitigging/asm.ovlStore \
      -evalues \
      -L /io/test/asm/unitigging/3-overlapErrorAdjustment/oea.files \
    > /io/test/asm/unitigging/3-overlapErrorAdjustment/oea.apply.err 2>&1

-- Finished on Sun Mar 13 18:49:08 2016 (lickety-split) with 372.2 GB free disk space
----------------------------------------
ERROR:
ERROR:  Failed with exit code 1.
ERROR:
================================================================================
Don't panic, but a mostly harmless error occurred and canu failed.

Disk space available:  372.2 GB

Last 50 lines of the relevant log file (/io/test/asm/unitigging/3-overlapErrorAdjustment/oea.apply.err):

loading evalues from '/io/test/asm/unitigging/3-overlapErrorAdjustment/0001.oea'
loading evalues from '/io/test/asm/unitigging/3-overlapErrorAdjustment/0001.oea' -- ID range 1-1393 with 58608 overlaps
Creating evalues file for 58654 overlaps....100.000%
memoryMappedFile()-- Couldn't mmap '/io/test/asm/unitigging/asm.ovlStore/evalues': Invalid argument

canu failed with 'failed to add error rates to overlap store'.
```

If instead I run:
```
cd $HOME
/io/canu-1.1/Linux-amd64/bin/canu -p asm -d asm -pacbio-raw /io/test/pacbio.filtered_subreads.fasta genomeSize=50k java=/io/test/jre1.8.0_73/bin/java minReadLength=500 stopOnReadQuality=false

-- Finished on Sun Mar 13 19:13:45 2016 (lickety-split) with 14.4 GB free disk space
----------------------------------------
-- Unitig sequences saved in '/home/asm/asm.*.fasta'.
-- Summary saved in '/home/asm/unitigging.html'.
```

So this isn't a Canu issue, the mmap command being used is correct. The issue is the docker container interaction with the outside world.
